src/main/java/org/jfrog/hudson/pipeline/types/NpmBuild.java

```diff
@@ -0,0 +1,131 @@
+
+public class NpmBuild implements Serializable {
+    private transient CpsScript cpsScript;
+    private NpmDeployer deployer = new NpmDeployer();
+    private NpmResolver resolver = new NpmResolver();
+    private String executablePath = "";
+
+    public NpmBuild() {
+    }
+
+    public void setCpsScript(CpsScript cpsScript) {
+        this.cpsScript = cpsScript;
+    }
+
+    @Whitelisted
+    public NpmDeployer getDeployer() {
+        return this.deployer;
+    }
+
+    @Whitelisted
+    public NpmResolver getResolver() {
+        return this.resolver;
+    }
+
+    @Whitelisted
+    public String getExecutablePath() {
+        return executablePath;
+    }
+
+    @Whitelisted
+    public void setExecutablePath(String executablePath) {
+        this.executablePath = executablePath;
+    }
+
+    @Whitelisted
+    public void install(Map<String, Object> args) {
+        // Throws CpsCallableInvocation - Must be the last line in this method
+        cpsScript.invokeMethod("artifactoryNpmInstall", prepareNpmStep(args));
+    }
+
+    @Whitelisted
+    public void publish(Map<String, Object> args) {
+        // Throws CpsCallableInvocation - Must be the last line in this method
+        cpsScript.invokeMethod("artifactoryNpmPublish", prepareNpmStep(args));
+    }
+
+    private Map<String, Object> prepareNpmStep(Map<String, Object> args) {
+        Set<String> npmArgumentsSet = args.keySet();
+        List<String> keysAsList = Arrays.asList("path", "args", "buildInfo");
+        if (!keysAsList.containsAll(npmArgumentsSet)) {
+            throw new IllegalArgumentException("Only the following arguments are allowed: " + keysAsList.toString());
+        }
+
+        deployer.setCpsScript(cpsScript);
+        Map<String, Object> stepVariables = getRunArguments((String) args.get("path"), (String) args.get("args"), (BuildInfo) args.get("buildInfo"));
+        appendBuildInfo(cpsScript, stepVariables);
+        return stepVariables;
+    }
+
+    @Whitelisted
+    public void resolver(Map<String, Object> resolverArguments) throws Exception {
+        Set<String> resolverArgumentsSet = resolverArguments.keySet();
+        List<String> keysAsList = Arrays.asList("repo", "server");
+        if (!keysAsList.containsAll(resolverArgumentsSet)) {
+            throw new IllegalArgumentException("Only the following arguments are allowed: " + keysAsList.toString());
+        }
+
+        // We don't want to handle the deserialization of the ArtifactoryServer.
+        // Instead we will remove it and later on set it on the deployer object.
+        Object server = resolverArguments.remove("server");
+        JSONObject json = new JSONObject();
+        json.putAll(resolverArguments);
+
+        final ObjectMapper mapper = new ObjectMapper();
+        mapper.readerForUpdating(resolver).readValue(json.toString());
+        if (server != null) {
+            resolver.setServer((ArtifactoryServer) server);
+        }
```

Comment:
The above code repeats in this class (NpmBuild), as well as in MavenBuild and GradleBuild.
Can we perhaps reuse it? Maybe by by putting it in a super class of the three classes?
Please note that the code is actually duplicated  in the *resolver* and *deployer* methods of the above classes.